### PR TITLE
Start moving webapp to Redux

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
Moves the user store, preference store and part of the team store over to use Redux.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-6212

### CODE REVIEW

File: webapp/components/channel_invite_modal.jsx [hunk 1]
```diff
@@ -53,7 +63,7 @@ export default class ChannelInviteModal extends React.Component {
         UserStore.addNotInChannelChangeListener(this.onChange);
         UserStore.addStatusesChangeListener(this.onStatusChange);
 
-        AsyncClient.getProfilesNotInChannel(this.props.channel.id, 0);
+        this.props.actions.getProfilesNotInChannel(TeamStore.getCurrentId(), this.props.channel.id, 0);
```
Comment (enahum) [inline webapp/components/channel_invite_modal.jsx]: again you should pass the currentTeamId as a prop or use the selector here

File: webapp/components/channel_invite_modal.jsx [hunk 2]
```diff
@@ -103,40 +115,24 @@ export default class ChannelInviteModal extends React.Component {
     }
 
     nextPage(page) {
-        AsyncClient.getProfilesNotInChannel(this.props.channel.id, (page + 1) * USERS_PER_PAGE, USERS_PER_PAGE);
+        this.props.actions.getProfilesNotInChannel(TeamStore.getCurrentId(), this.props.channel.id, (page + 1) * USERS_PER_PAGE, USERS_PER_PAGE);
```
Comment (enahum) [inline webapp/components/channel_invite_modal.jsx]: same

File: webapp/components/channel_invite_modal.jsx [hunk 3]
```diff
@@ -103,40 +115,24 @@ export default class ChannelInviteModal extends React.Component {
...
-                    {not_in_channel_id: this.props.channel.id},
-                    (users) => {
-                        if (searchTimeoutId !== this.searchTimeoutId) {
-                            return;
-                        }
-
-                        this.setState({search: true, users});
-                    }
-                );
+                searchUsers(term, TeamStore.getCurrentId(), {not_in_channel_id: this.props.channel.id});
```
Comment (enahum) [inline webapp/components/channel_invite_modal.jsx]: same

File: webapp/components/login/login_controller.jsx [hunk 4]
```diff
@@ -209,7 +209,7 @@ export default class LoginController extends React.Component {
     }
 
     finishSignin(team) {
-        GlobalActions.emitInitialLoad(
+        loadMe(
```
Comment (enahum) [inline webapp/components/login/login_controller.jsx]: maybe im not reading this correctly but it this is called after `weblogin` that in turns calls the redux action `login` then `loadMe` is not necessary as you are going to perform the same requests more than once

File: webapp/components/member_list_channel.jsx [hunk 5]
```diff
@@ -87,43 +89,30 @@ export default class MemberListChannel extends React.Component {
...
             return;
         }
 
         const searchTimeoutId = setTimeout(
             () => {
-                searchUsers(
-                    term,
-                    TeamStore.getCurrentId(),
-                    {},
+                searchUsers(term, TeamStore.getCurrentId(), {},
```
Comment (enahum) [inline webapp/components/member_list_channel.jsx]: remember this? hehe

File: webapp/components/member_list_team.jsx [hunk 6]
```diff
@@ -67,15 +69,15 @@ export default class MemberListTeam extends React.Component {
...
+    onChange() {
+        let users;
+        if (this.term) {
+            users = searchProfilesInCurrentTeam(store.getState(), this.term);
+        } else {
+            users = UserStore.getProfileListInTeam();
         }
 
-        this.setState({users: UserStore.getProfileListInTeam(), teamMembers: Object.assign([], TeamStore.getMembersInTeam())});
+        this.setState({users, teamMembers: Object.assign([], TeamStore.getMembersInTeam())});
```
Comment (enahum) [inline webapp/components/member_list_team.jsx]: selector for `membersInTeam`

File: webapp/components/more_direct_channels.jsx [hunk 7]
```diff
@@ -63,9 +75,9 @@ export default class MoreDirectChannels extends React.Component {
         UserStore.addStatusesChangeListener(this.onChange);
 
         if (this.listType === 'any') {
-            AsyncClient.getProfiles(0, USERS_PER_PAGE * 2);
+            this.props.actions.getProfiles(0, USERS_PER_PAGE * 2);
         } else {
-            AsyncClient.getProfilesInTeam(TeamStore.getCurrentId(), 0, USERS_PER_PAGE * 2);
+            this.props.actions.getProfilesInTeam(TeamStore.getCurrentId(), 0, USERS_PER_PAGE * 2);
```
Comment (enahum) [inline webapp/components/more_direct_channels.jsx]: currentTeamId as a prop?

File: webapp/containers/add_users_to_team.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import {connect} from 'react-redux';
+import {bindActionCreators} from 'redux';
+import {getProfilesNotInTeam} from 'mattermost-redux/actions/users';
+
+import AddUsersToTeam from 'components/add_users_to_team.jsx';
+
+function makeMapStateToProps() {
```
Comment (enahum) [inline webapp/containers/add_users_to_team.js]: why are we wrapping the mapStateToProps function in another function here? it doesn't seen necessary

File: webapp/containers/channel_invite_modal.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import {connect} from 'react-redux';
+import {bindActionCreators} from 'redux';
+import {getProfilesNotInChannel} from 'mattermost-redux/actions/users';
+
+import ChannelInviteModal from 'components/channel_invite_modal.jsx';
+
+function makeMapStateToProps() {
```
Comment (enahum) [inline webapp/containers/channel_invite_modal.js]: same here

File: webapp/containers/more_direct_channels.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import {connect} from 'react-redux';
+import {bindActionCreators} from 'redux';
+import {getProfiles, getProfilesInTeam} from 'mattermost-redux/actions/users';
+
+import MoreDirectChannels from 'components/more_direct_channels.jsx';
+
+function makeMapStateToProps() {
```
Comment (enahum) [inline webapp/containers/more_direct_channels.js]: and here

File: webapp/containers/popover_list_members.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import {connect} from 'react-redux';
+import {bindActionCreators} from 'redux';
+import {getProfilesInChannel} from 'mattermost-redux/actions/users';
+
+import PopoverListMembers from 'components/popover_list_members.jsx';
+
+function makeMapStateToProps() {
```
Comment (enahum) [inline webapp/containers/popover_list_members.js]: again

File: webapp/containers/team_members_dropdown.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import {connect} from 'react-redux';
+import {bindActionCreators} from 'redux';
+import {getUser} from 'mattermost-redux/actions/users';
+
+import TeamMembersDropdown from 'components/team_members_dropdown.jsx';
+
+function makeMapStateToProps() {
```
Comment (enahum) [inline webapp/containers/team_members_dropdown.js]: ok this is a pattern

File: webapp/package.json [hunk 13]
```diff
@@ -19,8 +19,10 @@
     "intl": "1.2.5",
     "jasny-bootstrap": "3.1.3",
     "jquery": "3.1.1",
+    "localforage": "1.5.0",
     "marked": "mattermost/marked#8f5902fff9bad793cd6c66e0c44002c9e79e1317",
     "match-at": "0.1.0",
+    "mattermost-redux": "mattermost/mattermost-redux#webapp",
```
Comment (enahum) [inline webapp/package.json]: interesting, why not master?

File: webapp/store/index.js [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    'profilesNotInTeam'
+];
+
+const teamSetTransform = [
+    'membersInTeam'
+];
+
+const setTransforms = [
+    ...usersSetTransform,
+    ...teamSetTransform
```
Comment (enahum) [inline webapp/store/index.js]: outdated!

File: webapp/store/index.js [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+                return state;
+            }
+
+            return outboundState;
+        }
+    );
+
+    const offlineOptions = {
+        persist: (store, options) => {
+            const persistor = persistStore(store, {storage: localForage, ...options}, () => {
```
Comment (enahum) [inline webapp/store/index.js]: are we going to persists the store in localstorage?

File: webapp/stores/redux_store.jsx [hunk 16]
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+// This is a temporary store while we are transitioning from Flux to Redux. This file exports
+// the configured Redux store for use by actions and selectors.
+
+import configureStore from 'store';
```
Comment (enahum) [inline webapp/stores/redux_store.jsx]: im guessing this is the same as `./store` right?

File: webapp/actions/user_actions.jsx [hunk 17]
```diff
@@ -19,6 +19,64 @@ import Client from 'client/web_client.jsx';
...
+                        referrer: '',
+                        search: '',
+                        title: '',
+                        url: ''
+                    },
+                    anonymousId: '00000000000000000000000000'
+                });
+            }
+
+            localStorage.setItem('currentUserId', UserStore.getCurrentId());
```
Comment (jwilander) [inline webapp/actions/user_actions.jsx]: This file will get deleted eventually anyways so it doesn't really matter

File: webapp/actions/user_actions.jsx [hunk 18]
```diff
@@ -19,6 +19,64 @@ import Client from 'client/web_client.jsx';
...
+
+            getMyTeams()(dispatch, getState).then(() => {
```
Comment (jwilander) [inline webapp/actions/user_actions.jsx]: Hmmm, for some reason I thought it wasn't. I'll remove this

File: webapp/actions/user_actions.jsx [hunk 19]
```diff
@@ -491,119 +470,69 @@ function onThemeSaved(teamId, theme, onSuccess) {
...
     );
 }
 
-export function autocompleteUsersInChannel(username, channelId, success, error) {
-    Client.autocompleteUsersInChannel(
-        username,
-        channelId,
+export function autocompleteUsersInChannel(username, channelId, success) {
+    const channel = ChannelStore.get(channelId);
+    const teamId = channel ? channel.team_id : TeamStore.getCurrentId();
```
Comment (jwilander) [inline webapp/actions/user_actions.jsx]: Like I said, doesn't really matter as these actions will get moved into redux directly

File: webapp/actions/user_actions.jsx [hunk 20]
```diff
@@ -907,7 +794,7 @@ export function uploadProfileImage(userPicture, success, error) {
     Client.uploadProfileImage(
         userPicture,
         () => {
-            AsyncClient.getMe();
+            getMe()(dispatch, getState);
```
Comment (jwilander) [inline webapp/actions/user_actions.jsx]: Are you thinking of loadMe?

File: webapp/store/index.js (hunk 15 above)
Comment (jwilander) [inline webapp/store/index.js]: Nope, using localForage we'll be using IndexedDB or WebSQL

**Comment by jwilander** [comment]
Regarding the majority of your comments:

No point in using the selectors over the flux stores in the action files, all those actions will be moved to the Redux repo anyways. As for the props vs using the selectors, those changes will come later when those components actually become pure.

File: webapp/components/login/login_controller.jsx (hunk 4 above)
Comment (jwilander) [inline webapp/components/login/login_controller.jsx]: Possibly, I didn't look at this one very closely

File: webapp/stores/redux_store.jsx (hunk 16 above)
Comment (jwilander) [inline webapp/stores/redux_store.jsx]: Not sure what you mean, it's pulling in webapp/store/index.js with this

File: webapp/package.json (hunk 13 above)
Comment (jwilander) [inline webapp/package.json]: Because it needs a whole bunch of stuff from my redux webapp branch :P Once that's merged I'll switch it over to master but in the mean time to get testing done it's pointed at the branch

File: webapp/containers/add_users_to_team.js (hunk 8 above)
Comment (jwilander) [inline webapp/containers/add_users_to_team.js]: I just copied what the RN files were doing such as https://github.com/mattermost/mattermost-mobile/blob/master/app/components/post/index.js#L19 but I guess I copied the wrong one, since that needs it for the factory selector. I'll switch them over

Comment (enahum) [inline webapp/containers/add_users_to_team.js]: did you forget about these?

Comment (jwilander) [inline webapp/containers/add_users_to_team.js]: Hold your horses, not done yet :P

File: webapp/store/index.js (hunk 14 above)
Comment (jwilander) [inline webapp/store/index.js]: Not sure what you mean, is there a new way to do this now? The RN repo looks like it's doing the same way unless I'm missing something

Comment (enahum) [inline webapp/store/index.js]:
```js
const usersSetTransform = [
    'profilesInChannel',
    'profilesNotInChannel',
    'profilesInTeam'
];

const teamSetTransform = [
    'membersInTeam'
];

const channelSetTransform = [
    'channelsInTeam'
];

const setTransforms = [
    ...usersSetTransform,
    ...teamSetTransform,
    ...channelSetTransform
];
```

Comment (jwilander) [inline webapp/store/index.js]: Ah that's not even merged into `master` yet, I'll update it in a later PR

**Comment by jasonblais** [comment]
@jwilander Tested the following. I can continue testing more later, but I think I found a few things that should keep you busy for a bit :)

 - promoting member to team admin, channel admin
 - posting messages, replies, edits, permalink view, search, recent mentions, pinned/flagged posts searches
 - pinning posts, flagging, 
 - checking statuses
 - adding users to a team and channel
 - creating DM/GM channels
 - updating account settings > general info
 - viewing/revoking sessions
 - clicking/checking random things on the Account Settings modal

1) Main Menu is missing options for a regular member and team admins

Observed: Missing fields
 - get team invite link [this seems to reproduce on `master` so I can create a separate ticket for this]
 - create team [team creation is enabled]
 - join another team
 - report a problem / help / download apps links

Note that these options are shown for a system admin. None of these options have been restricted to admins only from the System Console

2) After a refresh, all statuses appear offline.

image
^haven't used slash commands to manually set status. All accounts are online / had activity in the last 60 seconds.

3) Not receiving any desktop notifications in Chrome (only hear a 'ping' sound) nor in Firefox

Notification settings:
image

4) JS error after reloading a channel (or switching to a channel) containing messages

image

5) Replies don't work (hitting Enter in RHS does nothing + get JS errors)

image

6) JS error each time after typing a letter following an "@"

image

7) Recent mentions don't work [clicking the "@" sign does nothing]

8) Flagging a post doesn't work [clicking the flag icon does nothing]

9) Cannot open a DM with a person outside the team

 - go to System Console > Users and Teams and set "Enable users to open Direct Message channels with:" to "Any user on the Mattermost server"
image
 - create a team
 - click "More..." at the bottom of LHS to open "More DM" modal
 - observed: No one to choose from [but switching to a team with members let's me start a DM or GM with those people]
image

10) Random: I saw this exception in the console sometimes

image

11) viewing Account Settings > Security > Active Sessions opens a modal, but info never loads
image

12) viewing active sessions from Account Settings > Security opens a "help" dialog. But can't actually view or revoke them

image

13) When mobile push notifications aren't configured/enabled, help text for mobile push notifications should say "Disabled by System Administrator"

Observed [expanding the settings does say it's disabled; but it doesn't say so in the attached help text]:
image
image
^second screenshot is correct


Expected:
image

14) Selecting options in Account Settings > Notifications don't get saved (after saving, resets to default option)

15) Updating password gives "Invalid Date" as the time of last update [the password does update though]
image

16) Updating themes doesn't work (cannot save changes + JS error)

image

17) Updating font doesn't work
image

^I think the first error came from clicking outside the Account Settings modal

**Comment by jwilander** [comment]
@jasonblais Nice work! That was extremely helpful. Most of those issues arose from just a few cascading problems and I've pushed fixes for all of them

**Comment by jasonblais** [comment]
Tested the following on Win10 / Chrome + Firefox

 - confirmed previously reported issues are fixed
 - promoting member to team admin, channel admin
 - posting messages, replies, edits, permalink view, search, recent mentions, pinned/flagged posts searches
 - clicking "Load more messages"
 - pinning posts, flagging, 
 - checking statuses
 - adding users to a team and channel
 - creating DM/GM channels
 - updating account settings > general info
 - viewing/revoking sessions
 - clicking/checking random things on the Account Settings modal


General observation: I feel the initial load of some features is slower: E.g. my status is shown as offline for a split second before it changes to online. Also, sometimes modals load only partially until all information is displayed (again, the delay is small, but noticeable).

1) Pinning a post works but gives a JS error

image

2) Recent mentions search doesn't work (clicking "@" does nothing)

3) Replying to a post in flagged posts list throws an error

 - flag a post
 - click on the flag icon to display flagged posts list
 - click reply arrow
 - post a reply
 - observed:
image

4) Initially desktop notifications worked on Chrome, but then I started getting JS errors

image

5) When attempting to change a password, entering an incorrect "Current Password" gives wrong error message and JS errors

image

Expected: Error message `The "Current Password" you entered is incorrect. Please check that Caps Lock is off and try again.`

6) Revoking a session changes UI and causes a JS error before logs you out

**Comment by jasonblais** [comment]
Argh, sorry for delay, battling connection issues :/

----

Tested the following on Win10 / Chrome + Firefox

 - confirmed previously reported issues are fixed
 - promoting member to team admin, channel admin
 - posting messages, replies, edits, permalink view, search, recent mentions, pinned/flagged posts searches
 - clicking "Load more messages"
 - pinning posts, flagging, 
 - checking statuses
 - adding users to a team and channel
 - creating DM/GM channels
 - updating account settings > general info
 - viewing/revoking sessions
 - clicking/checking random things on the Account Settings modal

I didn't test settings around System Console (e.g. user permissions) nor logging in/out of various account types, nor WebRTC yet

----

General observation: I feel the initial load of some features is slower: E.g. my status is shown as offline for a split second before it changes to online. Also, sometimes modals load only partially until all information is displayed (again, the delay is small, but noticeable).

1) Pinning a post works but gives a JS error

image

2) Recent mentions search doesn't work (clicking "@" does nothing)

3) Replying to a post in flagged posts list throws an error

 - flag a post
 - click on the flag icon to display flagged posts list
 - click reply arrow
 - post a reply
 - observed:
image

4) Initially desktop notifications worked on Chrome, but then I started getting JS errors.

They started working again after logout/login.

image

5) When attempting to change a password, entering an incorrect "Current Password" gives wrong error message and JS errors

image

Expected: Error message `The "Current Password" you entered is incorrect. Please check that Caps Lock is off and try again.`

6) Revoking a session changes UI and causes a JS error before logs you out

image

7) Typing in a channel containing replies throws JS errors (doesn't reproduce consistently, reproduces more frequently for text containing `@` or `~` that opens an autocomplete)

image

8) The spinmint site doesn't load for me on iOS Safari browser (can't get to the login page, just loads forever). Not sure if it's on my device (due to connection issues), but other MM sites seemed to load fine

9) Emoji sprites on the picker aren't visible

image

10)

**Comment by jwilander** [comment]
Alright thanks @jasonblais, I think I fixed most of the issues:

1) I'm not sure what was causing this but I can't reproduce and the code in that JS error isn't related to anything I touched. Not sure if you found some corner case bug but I don't think it's related to my changes
2/3/4/7) Were all caused by the same issue, which was the the current user was getting overwritten by autocomplete (so your notification preferences would get wiped if you typed '@')
5) Fixed after I saw your accidental post ;)
8) Not sure why it's not working, going to charge my iPod to try it out
9) I didn't change anything related to the emoji picker and it works fine for me, maybe it's your connection issues?

As for the general slowness you observed, the only thing I can think of is that the initial load might be a bit slower since Redux does that as different requests. That should only affect the loading before anything at all renders in the UI though, and won't affect statuses or modals at all. For me it doesn't seem any slower but maybe something to watch out for

**Comment by jwilander** [comment]
Figured out and fixed it on iOS Safari. The reason that didn't work is because javascript is dumb :/

**Comment by jasonblais** [comment]
1. Oh right, this is an edge case I found earlier on `master`. Help wanted ticket here, no need to fix in this PR: https://github.com/mattermost/platform/issues/6145
9. True, might be connection-related, will re-test on the new spinmint once it's up

Thanks for the others!

**Comment by jwilander** [comment]
Hmmm seems something is broken, let me check

**Comment by jwilander** [comment]
Ok it was a change to EN's redux-offline fork that broke it for production. I've fixed the issue and started the rebuild

**Comment by jwilander** [comment]
@jasonblais Alright, I think it's finally ready

**Comment by jasonblais** [comment]
Confirmed above fixes are done, except 

1 - has a help wanted ticket: #6145
9 - working, probably caused by slow connection then
10 - There's a delay between revoking a session and actually getting the session revoked (~5 seconds). On `master` the action happens immediately (< 1s)



image

11 - I can't seem to connect to an AD/LDAP server. Tried configuring based on 
 - https://github.com/mattermost/platform-private/blob/master/config/config_pre_release.json
 - https://github.com/mattermost/platform-private/blob/master/config/config_ci_linux_mysql.json

image

WebRTC: Works on Firefox, needed a couple of tries on Chrome but I think it's related to known connection issues on WebRTC.

No other issues found so far

**Comment by jwilander** [comment]
10 - I believe the delay is based on when your next API request goes out, so sometimes it will be immediate if you click on something that causes a request, other times it will be a few seconds until your browser makes another request

11 - The LDAP server is behind the VPN and the spinmints aren't so they won't be able to connect to it, I think this one is fine to test once it's on CI

**Comment by jasonblais** [comment]
@jwilander Any other areas that come to mind that I should test?

There might still be some edge cases, but haven't found any new ones so far..

If we get it soaking on pre-release, we'll hopefully catch the remaining bugs before code complete + LI or LM will do some testing before closing the Jira ticket

**Comment by jwilander** [comment]
Yeah, let's get it dev reviewed and then soak it on pre-release to find the
corner cases
